substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+                            if (mallocedPacket) {
+                                //   871                      free(fullPacket);
+                                LibC.free(fullPacket);
+                                /* so that the outter try/finally does not attempt to double-free() */
+                                mallocedPacket = false;
+                            }
+                            //   874                  return;
+                            return;
+                        }
+                    }
+                }
+
+                //   878          len = SOCKADDR_LEN;
+                len_Pointer.write(JavaNetNetUtilMD.SOCKADDR_LEN());
+
+                //   879          n = NET_RecvFrom(fd, fullPacket, packetBufferLen, 0,
+                //   880                           (struct sockaddr *)&remote_addr, &len);
+
+                n = JavaNetNetUtilMD.NET_RecvFrom(fd, fullPacket, packetBufferLen, 0, remote_addr, len_Pointer);
+                //   881          /* truncate the data if the packet's length is too small */
+                //   882          if (n > packetBufferLen) {
+                if (n > packetBufferLen) {
+                    //   883              n = packetBufferLen;
+                    n = packetBufferLen;
+                }
+
+                //   885          if (n == JVM_IO_ERR) {
+                if (n == Target_jvm.JVM_IO_ERR()) {
+                    //   886              (*env)->SetIntField(env, packet, dp_offsetID, 0);
+                    Util_java_net_DatagramPacket.as_Target_java_net_DatagramPacket(packet).offset = 0;
+                    //   887              (*env)->SetIntField(env, packet, dp_lengthID, 0);
+                    Util_java_net_DatagramPacket.as_Target_java_net_DatagramPacket(packet).length = 0;
+                    if (Errno.errno() == Errno.ECONNREFUSED()) {
+                        //   888              if (errno == ECONNREFUSED) {
+                        throw new PortUnreachableException("ICMP Port Unreachable");
+                        //   889                  JNU_ThrowByName(env, JNU_JAVANETPKG "PortUnreachableException",
+                        //   890                                  "ICMP Port Unreachable");
+                        //   891              } else {
+                    } else {
+                        //   892                  if (errno == EBADF) {
+                        if (Errno.errno() == Errno.EBADF()) {
+                            //   893                       JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException", "Socket closed");
+                            throw new SocketException("Socket closed");
+                        } else {
+                            //   895                       NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException", "Receive failed");
+                            throw new SocketException("Receive failed");
```

Comment:
Missing the `PosixUtils.lastErrorString` part of the translation.  Hmm, and now that I look, I missed commenting on several others that are missing.